### DISCUSSION TITLE
Usage of EventToCommandBehavior with custom UserControl

### DISCUSSION THREAD

**Discussion by xaberue** [CommunityToolkit/Maui#1880] [Q&A]
Hi Community!

I am trying to bind an event raised from a UserControl that is reused in other places of my .NET MAUI app.

This UserControl, brom the code behind raises and event called FilterChanged.

```csharp

///

public static readonly BindableProperty FilterChangedProperty = 
    BindableProperty.Create(nameof(FilterChanged), typeof(EventHandler<ContactTypeSelectionArgs>), typeof(ContactTypeSelectorControl), defaultBindingMode: BindingMode.OneWay, defaultValueCreator: default);


public EventHandler<ContactTypeSelectionArgs> FilterChanged
{
    get => (EventHandler<ContactTypeSelectionArgs>)GetValue(FilterChangedProperty);
    set => SetValue(FilterChangedProperty, value);
}

///
```

Now, trying to capture this event from one page, I found [this](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/maui/behaviors/event-to-command-behavior) behaviour that on paper, fits my requirements.

However, once I try to use it, I'm never able to start the app, what am I doing wrong? How is it supposed to be used?

XAML:
```xaml
<controls:ContactTypeSelectorControl ContactTypes="{Binding ContactTypes}" >
    <controls:ContactTypeSelectorControl.Behaviors>
        <toolkit:EventToCommandBehavior x:TypeArguments="controls:ContactTypeSelectionArgs" EventName="FilterChanged" Command="{Binding FilterChangedCommand}" />
    </controls:ContactTypeSelectorControl.Behaviors>
</controls:ContactTypeSelectorControl>
```

ViewModel:
```csharp
[RelayCommand]
private async Task OnFilterChangedAsync(ContactTypeSelectionArgs criteria)
{
    //TODO

    await Task.CompletedTask;
}
```

Error
> System.Reflection.TargetInvocationException: 'Exception has been thrown by the target of an invocation.' :  ArgumentException: EventToCommandBehavior: Couldn't resolve the event. Arg_ParamName_Name

I've tried several ways, including using AncestorType and always having the same problem. 

Additionally, how can I capture the param from the event into the Command Handler? Do I need to specify a CommandParameter for this?

Many thanks in advance!

**Comment by bijington** [comment]
Upon first glance your event definition looks rather strange - you don't tend to make those a BindableProperty.

What happens if you try

```csharp 
public event EventHandler<ContactTypeSelectionArgs> FilterChanged;
```

Also for your question on getting the event args in as a parameter, this should just happen for you and you shouldn't need to set anything else up.

**Comment by xaberue** [comment]
Hi @bijington!

Many thanks! that was everything!

I was carrying this code from previous samples (in WPF) and tests, it was simpler, and I was struggling with an invalid event definition...

With your proposed change, everything now works as expected, and I' receiving the param as well without any problem.

```csharp
//

public event EventHandler<ContactTypeSelectionArgs> FilterChanged;

//

public void OnTapGestureRecognizerTapped(object sender, TappedEventArgs args)
{
    var selection = (ContactTypeElement)args.Parameter!;

    FilterChanged?.Invoke(this, new ContactTypeSelectionArgs { TypeSelected = (ContactTypeEnum)selection.Id });
}

//
```

```xaml
<controls:ContactTypeSelectorControl ContactTypes="{Binding ContactTypes}" Margin="10" x:Name="contactType">
    <controls:ContactTypeSelectorControl.Behaviors>
        <toolkit:EventToCommandBehavior 
                x:TypeArguments="controls:ContactTypeSelectionArgs" 
                EventName="FilterChanged"
                Command="{Binding FilterChangedCommand}" />
    </controls:ContactTypeSelectorControl.Behaviors>
</controls:ContactTypeSelectorControl>
```

![image](https://github.com/CommunityToolkit/Maui/assets/55638917/0c49f29d-7bb3-4401-94bd-53edc329b58a)

Thanks again!

**Reply by bijington** [reply]
That's great to hear. Thanks for confirming
